src/java.base/share/classes/java/io/IO.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package java.io;
+
+import jdk.internal.javac.PreviewFeature;
+
+/**
+ * A collection of static convenience methods that provide access to
+ * {@linkplain System#console() system console} for implicitly declared classes.
+ *
+ * <p> Each of this class' methods throws {@link IOError} if the system console
+ * is {@code null}; otherwise, the effect is as if a similarly-named method
+ * had been called on that console.
+ *
+ * <p> Output from methods in this class uses the character set of the system
+ * console as specified by {@link Console#charset}.
+ *
+ * @since 23
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.IMPLICIT_CLASSES)
+public class IO {
+
+    private IO() {
+        throw new Error("no instances");
```

Comment:
Is this necessary?

Reply:
No, it's not necessary.

My guess is that throwing an unconditional error (usually with a snarky message) from the inside the sole private constructor is a historical idiom to deter deep reflectors with their `setAccessible` and what not. Maybe that idiom is outdated with modules and recent efforts on cranking up the [integrity](https://openjdk.org/jeps/8305968), that I don't know.

FWIW, here's a snippet[^*] from "Item 4: Enforce noninstantiability with a private constructor" you referred to in your earlier comment:

> A default constructor is generated only if a class contains no explicit constructors, so a class can be made noninstantiable by including a private constructor:
>
>        // Noninstantiable utility class
>        public class UtilityClass {
>            // Suppress default constructor for noninstantiability
>            private UtilityClass() {
>                throw new AssertionError();
>            }
>            ...  // Remainder omitted
>        }
>
> Because the explicit constructor is private, it is inaccessible outside of the class. The `AssertionError` isn’t strictly required, but it provides insurance in case the constructor is accidentally invoked from within the class. It guarantees the class will never be instantiated under any circumstances.

[^*]: I created that snippet from the second edition, whose page 19 is conveniently publicly available here: https://www.google.ie/books/edition/Effective_Java/ka2VUBqHiWkC?hl=en&gbpv=1&dq=Item+4:+Enforce+noninstantiability+with+a+private+constructor&pg=PA19&printsec=frontcover